### ISSUE TITLE
How to visualize an inference result?

### ISSUE THREAD

**Issue by slavic-codeman** [zhuduowang/Change3D#12]
Hi! I have written an inference code based on your validation process, but the visualization is not correct:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/5cd7beda-a53a-4513-8873-62a3e0b13839" />

My code follows your validation code, and I notice that you use 0.5 for threshholding to calculate F1 score. My code is:
```
# Copyright (c) Duowang Zhu.
# All rights reserved.

import os
import sys
import time
import torch
import torch.nn.functional as F
import torch.backends.cudnn as cudnn
import numpy as np
from tqdm import tqdm
from skimage import io
# Insert current path for local module imports
sys.path.insert(0, '.')
from PIL import Image
import data.dataset as RSDataset
import data.transforms as RSTransforms
from utils.metric_tool import ConfuseMatrixMeter
from model.trainer import Trainer
from model.utils import BCEDiceLoss
import cv2

class ArgsObject:
    """将字典转换为具有属性的对象"""
    def __init__(self, config_dict):
        for key, value in config_dict.items():
            setattr(self, key, value)

def load_image(pre_img_path,post_img_path,label_path,transform):
    pre_image = io.imread(pre_img_path)
    post_image = io.imread(post_img_path)
    label = io.imread(label_path, as_gray=True)
    # Error handling for image loading failures
    if pre_image is None:
        raise IOError(f"Failed to load pre-change image: {pre_img_path}")
    if post_image is None:
        raise IOError(f"Failed to load post-change image: {post_img_path}")
    if label is None:
        raise IOError(f"Failed to load ground truth mask: {label_path}")
        
    # Concatenate pre and post images along the channel dimension
    # This creates a 6-channel image (BGR-BGR)
    img = np.concatenate((pre_image, post_image), axis=2)

    # Apply transforms if specified
    if transform:
        img, label = transform(img, label)
    img=img.unsqueeze(0)
   
    return img, label


def infer(model,pre_img_path,post_img_path,label_img_path,transform,device='cuda'):
    model.eval()
    img,label=load_image(pre_img_path,post_img_path,label_img_path,transform)
    pre_img = img[:, 0:3].to(device).float()
    post_img = img[:, 3:6].to(device).float()
      

            # Forward pass
    output = model.update_bcd(pre_img, post_img)
    pred = torch.where(
            output > 0.8,
            torch.ones_like(output),
            torch.zeros_like(output)
        ).long()
    return pred,label


def visualize(pred,path):
    # 去掉多余的维度，获取单一通道的 2D 图像
    if pred.dim() == 3:  # 如果是 [1, H, W]
        pred_np = pred.squeeze(0).cpu().numpy()
    else:  # 如果是 [H, W]
        pred_np = pred.cpu().numpy()

    # 转换为0-255范围的uint8
    pred_np = (pred_np * 255).astype(np.uint8)

    # 创建PIL图像并保存
    img = Image.fromarray(pred_np)
    img.save(path)

def read_gt(label_path):
   # label_path = "/network_space/server127_2/shared/tangjq/Change3D/CLCD/test/label/00480.png"

    # 1. 使用OpenCV读取图像（灰度模式）
    mask_np = cv2.imread(label_path, cv2.IMREAD_GRAYSCALE)

    # 2. 转换为PyTorch tensor并添加通道维度
    mask_tensor = torch.from_numpy(mask_np).unsqueeze(0)  # (1, H, W)

    # 3. 归一化到0-1范围
    mask_tensor = mask_tensor.float() / 255.0

    return mask_tensor
def main():
    """Main validation routine."""
    # Configuration (directly set variables instead of command line args)
    config = {
        'dataset': "CLCD",
        'file_root': "/network_space/server127_2/shared/tangjq/Change3D/CLCD",
        'in_height': 256,
        'in_width': 256,
        'num_perception_frame': 1,
        'num_class': 1,
        'batch_size': 16,
        'num_workers': 4,
        'pretrained': '/network_space/server127_2/shared/tangjq/Change3D/X3D_L.pyth',
        'model_checkpoint': '/network_space/server127_2/shared/tangjq/Change3D/exp/CLCD_iter_80000_lr_0.0002/best_model.pth',  # Path to trained model
        'gpu_id': 0,
        # 添加其他可能需要的参数
        'max_steps': 80000,
        'lr': 2e-4,
        'lr_mode': 'poly',
        'step_loss': 100,
        'save_dir': './exp',
        'resume': None,
        'log_file': 'train_val_log.txt'
    }
    
    # 将字典转换为对象，以兼容原来的 args.attribute 调用方式
    args = ArgsObject(config)
    
    # Set GPU device
    os.environ['CUDA_VISIBLE_DEVICES'] = str(args.gpu_id)
    device = 'cuda' if torch.cuda.is_available() else 'cpu'
    
    # Enable CUDA optimizations
    if device == 'cuda':
        cudnn.benchmark = True
    
    # Initialize model - 现在使用 args 对象
    model = Trainer(args).to(device).float()
    
    # Load trained model weights
    if os.path.exists(args.model_checkpoint):
        state_dict = torch.load(args.model_checkpoint, map_location=device)
        model.load_state_dict(state_dict)
        print(f"Loaded model from {args.model_checkpoint}")
    else:
        print(f"Warning: Model checkpoint not found at {args.model_checkpoint}")
        print("Running validation with randomly initialized weights.")
    
    # Create validation transform - 也需要传递 args 对象
    _, val_transform = RSTransforms.BCDTransforms.get_transform_pipelines(args)
    
    pre_img_path="/network_space/server127_2/shared/tangjq/Change3D/CLCD/train/t1/00000.png"
    post_img_path="/network_space/server127_2/shared/tangjq/Change3D/CLCD/train/t2/00000.png"
    label_img_path="/network_space/server127_2/shared/tangjq/Change3D/CLCD/train/label/00000.png"

    pred,label=infer(model,pre_img_path,post_img_path,label_img_path,val_transform)
    pred,label=pred[0].detach().cpu(),label[0].detach().cpu()
    visualize(pred,"pred.png")
    visualize(label,"label.png")
    #torch.save(output,"/network_space/server127_2/shared/tangjq/open-cd/opencd/apis/output/output.pt")
if __name__ == '__main__':
    main()
```
 The GT visualization is correct, and the model has been trained on CLCD.  Is there any problem with my code, and can you release your visualization code? Thanks!

**Comment by slavic-codeman** [comment]
I try different threshholding, it doesn't work.

**Comment by zhuduowang** [comment]
<img width="707" height="275" alt="Image" src="https://github.com/user-attachments/assets/fbce5d5b-7887-4c79-817f-7a3351a4a7a0" />

If the output has been passed through a sigmoid, please use 0.5 as the threshold.

**Comment by slavic-codeman** [comment]
Thank you! I find out that my pretrained weigths has been accidentally covered with untrained weights. Now the code can work. The visualization is correct!

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/cd3c3ac1-6840-4dda-8f66-bf6fdab4e93c" />
